### DISCUSSION TITLE
Combining pyramid map tile sets with ImageMagick

### DISCUSSION THREAD

**Discussion by stefankinell** [ImageMagick/ImageMagick#3921] [Help]
Hi,
I just started to learn how to use ImageMagick and found my way here because of a challenge I have. 

Task / question:
Is there a way to script ImageMagick to merge images that have the same name and folder structure in two parallel sets of base folders, in an easy way ;-)

The most fundamental task of overlaying two images I learned to do with the  composite command. But it is the scripting part I am not so familiar  with.

Background
I'm creating a map to be used on the web. The map tool used is leafletjs which uses tiles (.png-files). The files are stored in a structure of [z]/[x]/[y].png

To manage the task of building the map, I have to create one part of the map at the time. Creating multiple sets of tiles. I am looking for a way to merge, or overlay, the images that are on the borders of two parts of the map. 

As an example. In the map area A019 I have the image 5/7/3.png
![v](https://user-images.githubusercontent.com/59049892/125186683-16916d80-e22c-11eb-9fc0-2c1b2a1fb788.png)

In the area A020 i have an image in the same folder location 5/7/3.png
![h2](https://user-images.githubusercontent.com/59049892/125186699-2b6e0100-e22c-11eb-990f-f2516ed9a593.png)

They are two halfs of what should be in the end.
![final](https://user-images.githubusercontent.com/59049892/125186712-42acee80-e22c-11eb-9037-d0fe7e44918a.png)

Any ideas of how to accomplish this would be appreciated. (and since I use quite a specific tool to create my maps, that output the tiles in separate folders. The solution is not to create the tiles in a new way, I simply want to see if I can solve the merging of overlapping images in an automated way ;-) )

Thanks
Stefan

**Comment by snibgo** [comment]
What version of IM, on what platform?

What are the fully-qualified pathnames of those two files? (Fully-qualified means with absolute directory names.)

I suppose the IM command will be something like:
```
magick in1.png in2.png +append out.png
```
... where the three filenames are fully-qualified. The solution might involve a shell script, to setup the paths, but I'm unsure what the script should do overall.

**Reply by stefankinell** [reply]
Hi, thanks for your reply

Valid questions, I obviously missed some imput to the question :-)

It's intended to run on Windows. I just downloaded ImageMagick so the version should be  7.1.0-2.

I believe I have sorted our the command to combine the two images. The command
`magick composite D:\Export\D002\6\16\1.png D:\Export\Master\6\16\1.png D:\Export\D002\6\16\1.png`
did the trick of combining the two images and writing back the combined image to the location of my first image.

What I am trying to solve is how I do this recursively on a number of folders and images.

In the example above, z=6, x=16 and y=1

When z1 = z2 so to say, and x1 = x2 and y1 = y3 - then I want  to execute the ImageMagick command.

**Reply by fmw42** [reply]
Do the PNG files contain virtual canvas offsets to tell exactly where each time is offset from the top left corner of the whole map? If so, then they can be combined with a simple 

`magick image*.png -flatten result.png`

**Comment by snibgo** [comment]
Okay, you have an IM command. However, instead of "magick composite", I suggest you use "magick" with the "-composite" option. Instead of:
```
magick composite in1.png in2.png out.png
```
... you would have:
```
magick in2.png in1.png -composite out.png
```
Note that the order of the input files is reversed.

> What I am trying to solve is how I do this recursively on a number of folders and images.

This suggests you want to loop through subdirectories. If they are numbers, you could use "for /L" in Windows. For example (Windows BAT syntax):
```
for /L %%Z in (0,1,5) do (
  for /L %%X in (0,1,5) do (
    for /L %%Y in (0,1,5) do (
      echo %%Z\%%X\%%Y
    )
  )
)
```
I have a simple "echo" command. You would use "magick", of course.

**Reply by stefankinell** [reply]
I now manged to write a powershell script where I can compare the files in one set of folders with a second and then apply the Magic-command I want on each of the files that exists in both. I'm still working on performance by breaking the script down into loops within loops so I compare fewer files from old source and new source, which speeds up the process a lot. So good idea with looping through sub folders.

Once I'm done with a stable script I'll post back and place the code on github for others to use.

Now, for the ImageMagick command - what is the difference between 
`Magick composite ini1.png ini2.png out.png`
and
`magick in2.png in1.png -composite out.png`

Will they do the same and why then change my code? is one more efficient than the other?

I've also played around with removing white in the images, making them transparent instead, so it will work to overlay. But I'm not happy with the quality of the images. Is there a way to just ignore the white, treat it as background or something, so that part of the processes can be skipped? Is that what are refereed to as -alpha?

**Reply by fmw42** [reply]
The two commands are equivalent. But the second is much more flexible and uses the more current IM 7 syntax. It permits other (more complicated) processes to be added to the same command line. The first command would not allow that. In that way it would be more efficient and not have to do extra writes and reads for multiple command lines.

**Reply by stefankinell** [reply]
Perfect, that sounds good, Perhaps I can then make the script even more efficient by writing best way and combining two commands in one go.

I should probably close this topic since I got the help I asked for in the first place, and create a second topic for the actual composite command and the two images with white background.

Thanks @fmw42 and @snibgo

**Reply by fmw42** [reply]
I am not sure what you are trying to do with the white background. I thought you had non-overlapping tiles.  Can you post and example image with your question about the white.  

Here is a link to an example of doing multiple things in one command line using parenthesis processing.  https://legacy.imagemagick.org/Usage/files/#write

**Reply by stefankinell** [reply]
I see the images I used in the first example shows the images as transparent already. Here's another example, more true to reality.

I like to merge/overlay whatever terms one should use the following example pictures:
![21-2](https://user-images.githubusercontent.com/59049892/125336080-9cf1a080-e34d-11eb-95f3-13922d0ac710.png)

and
![21-1](https://user-images.githubusercontent.com/59049892/125336092-a0852780-e34d-11eb-99fa-cf4e08868b50.png)

Where there is white in one image, there is a drawing on the other. When setting the white to transparent it seems I loose some information/degrade the image since there are white spots in the actual map part as well. If possible I would therefore like a command where imagemagick just ignores the white, and merges the two images and keeps the background white.

**Reply by fmw42** [reply]
I am not sure I understand. Is this what you want?

`magick 125336080-9cf1a080-e34d-11eb-95f3-13922d0ac710.png  125336092-a0852780-e34d-11eb-99fa-cf4e08868b50.png -fuzz 35% -trim +repage -append result.png`


![result](https://user-images.githubusercontent.com/13458218/125339012-c2f35280-e305-11eb-9b41-81db5310f9ca.png)

**Reply by stefankinell** [reply]
That command seems to change the ratio of the image, so I get a resulting 256x258 pixles images instead of the intended. And even worse, when there are big white areas in between, the image considerable smaller. The intention was not to remove the white areas, instead of overlaying the other colors, in the same place, with the same aspect ratio so to say.

**Comment by fmw42** [comment]
So do you want this?

`magick 125336080-9cf1a080-e34d-11eb-95f3-13922d0ac710.png -append result.png`


![result](https://user-images.githubusercontent.com/13458218/125341821-ecfa4400-e308-11eb-8909-4a682f5551a1.png)

**Reply by stefankinell** [reply]
No, that seems tourist have added one image below the other. And now I see that they upper part of the map is attached below.

both images are 256x256 pixels. The end result should be a 256x256 pixel image. They should just be added on top of each other, just like the composite command does. But I’m not so fond of what the transparency function does to the image, which right now is my go to solution for making the composite command work with the white sections. I just want the composite command to ignore the white, if possible.

**Comment by snibgo** [comment]
The images are such poor quality that I'm not sure which way round they should be.

The "white" pixels are far from white. The boundary seems strange.

```
magick ^
  125336080-9cf1a080-e34d-11eb-95f3-13922d0ac710.png ^
  ( 125336092-a0852780-e34d-11eb-99fa-cf4e08868b50.png -bordercolor white -border 1 -fuzz 30%% -trim +repage ) ^
  -composite ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/125351450-01b1f880-e358-11eb-9c53-618775db17ca.png)

You could use transparency instead of trimming, but the result will still be poor.

**Comment by fmw42** [comment]
Which is the top image and which is the bottom image?

**Reply by fmw42** [reply]
Do they need a gap between them?

**Reply by fmw42** [reply]
Is this what you want?

`magick 125336092-a0852780-e34d-11eb-99fa-cf4e08868b50.png 125336080-9cf1a080-e34d-11eb-95f3-13922d0ac710.png -fuzz 45% -trim +repage -append result2.png`


![result2](https://user-images.githubusercontent.com/13458218/125362814-65223300-e324-11eb-9d73-599bf32cfb10.png)

**Reply by stefankinell** [reply]
Unfortunately that does not work very well. It is not allways that the tiles represent the full final picture in two. Let's use another example. These two images should be combined:
![4341](https://user-images.githubusercontent.com/59049892/125406624-52a60900-e3b9-11eb-944a-8d4b4daa2165.png)

and

![4343](https://user-images.githubusercontent.com/59049892/125406645-56399000-e3b9-11eb-9c38-8312badd67ef.png)

With your example above the result will be

![result2](https://user-images.githubusercontent.com/59049892/125406687-63ef1580-e3b9-11eb-8e82-2868f8f21e2b.png)

When the wanted outcome is

![result4](https://user-images.githubusercontent.com/59049892/125407219-f394c400-e3b9-11eb-8194-77509865ae31.png)

**Comment by snibgo** [comment]
My command seems to give the desired outcome. Note that this trims one image, not both.
```
magick ^
  125406645-56399000-e3b9-11eb-9c38-8312badd67ef.png ^
  ( 125406624-52a60900-e3b9-11eb-944a-8d4b4daa2165.png -bordercolor white -border 1 -fuzz 30%% -trim +repage ) ^
  -composite ^
  x.png
```

**Reply by stefankinell** [reply]
Thanks,
I can not get the parenthesis to work, so the command does not execute for me. I have to look into why this is happening because the same thing applied yesterday.

**Reply by fmw42** [reply]
Be sure you have spaces on each side of the parentheses. Also be sure there are no spaces after the ^ at the end of the lines and that each ^ ends a line. If not in a .bat file, then use single %.

(If on Unix and not Windows, then put \ before each parenthesis and end the lines with \ rather than ^).

**Reply by stefankinell** [reply]
I get `Unexpected token '^' in expression or statement.` for the line with the parenthesis.

I still have to consider if this is working, triming one image and making it smaller. Then I always have to know what pyramid is left and right, top or below. And that image will always need to stay at one side. I can not then merge two images that are not going all the way out to the edge.

With all this, perhaps my original idea with transparency still is the right way to go. And perhaps add white background if needed when the merge is done.

**Reply by fmw42** [reply]
Are you on DOS Window (rather than Windows 10 Unix)?  Does the ^ end the line (with no spaces after it) and do you have the next part on a separate line of text?

**Comment by snibgo** [comment]
My command is for Windows BAT syntax. For Windows CMD syntax (ie console), change the double %% to single %.

For bash, change the double %% to single % and escape the parentheses \\( and \\), and change line-endings from caret ^ to backslash \\.

I know nothing about Powershell syntax.
